openjdk/src/test/java/org/conscrypt/NativeCryptoTest.java

```diff
@@ -111,6 +111,8 @@ public class NativeCryptoTest {
     private static byte[] CHANNEL_ID;
     private static Method m_Platform_getFileDescriptor;
 
+    private static RSAPrivateCrtKey TEST_RSA_KEY;
```

Comment:
nit: Might as well make this an instance field (i.e. not static) and initialise it here rather than making getTestRsaKey() synchronized.   We're still on Junit 4 which is non-threaded anyway and when (if) we move to Junit 5, such field initialization is still thread safe.

Reply:
Ah right that locking isn't right. (I just copied it from elsewhere in the file.)

Does junit make a new instance per individual test or are they shared more broadly? (If it's per test, we don't gain anything from caching the key anyway.)

Reply:
Single instance for all tests[1]... So fields are shared between all the test methods and you do win by creating a single key.  And so the `static` fields aren't much of a win over instance fields except for primitive values / constants which can be compile-time initialised.

If you do need to do work for each test, you can annotate a `@Before` and/or `@After` method.

[1]  Or in the case of sharding, a subset of them. But then each shard gets its own test class object so there are still no concurrency issues.

Reply:
Ah okay I see.

Although looking  further, this file has two different ways of doing this, and this will introduce a third:

1. `m_Platform_getFileDescriptor` is initialized without locking in an `@BeforeClass` static method
2. All the others are initialized in a static, `synchronized` method like `initCerts`
3. (New one proposed here) Like 2, but with an instance method, this time without synchronization because it's not needed.

Is (3) actually the ideal pattern though? In so far as you can have multiple instances, seems we'd rather only do that work once. Having to lock things is tedious, but maybe we should just use `@BeforeClass`. Or is there a reason not to prefer that pattern?

Happy to just take a pass and align all of them on one pattern, unless there's a reason to use different ones for different cases? Seems the first might actually be ideal actually. I could just rename `getPlatformMethods` to `init` and do it all in one go.

WDYT?

> Ah right that locking isn't right. (I just copied it from elsewhere in the file.)

Correction, I think the locking for the old thing is fine? It's `static synchronized`, so it's presumably using a per-class lock.

Reply:
Sorry! This one fell off the radar...

Also I was incorrect in my previous explanation... I forgot the Junit actually creates a new instance for every test method, so whilst (3) is generally preferred, for expensive stuff (like this) in a class with lots of test methods static fields might be better.

Again with this you can instantiate the field directly, *except* you can't throw throw exceptions, so you might need to use `@BeforeClass`.

OTOH if you want to land this change as-is, it's perfectly safe and performant so feel free and I'll add this class to the list that need some general tidying up.  :)

Reply:
Moved everything into a `@BeforeClass`. Was that the pattern you had in mind or something else?

Reply:
Awesome, thanks!

Last CI run failed with SIGSEGV, which is surprising, _maybe_ transient... Trying again.

Reply:
Not transient but unrelated...